plugins/event_source/url_check.py

```diff
@@ -33,21 +32,33 @@ async def main(queue: asyncio.Queue, args: Dict[str, Any]):
         return
 
     while True:
-        async with aiohttp.ClientSession() as session:
-            for url in urls:
-                async with session.get(url) as resp:
-                    await queue.put(
-                        dict(
-                            url_check=dict(
-                                url=url,
-                                status="up" if resp.status == 200 else "down",
-                                status_code=resp.status,
+        try:
+            async with aiohttp.ClientSession() as session:
+                for url in urls:
+                    async with session.get(url) as resp:
+                        await queue.put(
+                            dict(
+                                url_check=dict(
+                                    url=url,
+                                    status="up" if resp.status == 200 else "down",
+                                    status_code=resp.status,
+                                )
                             )
                         )
+                        print(resp.status)
+
+        except aiohttp.ClientError as e:
+            await queue.put(
+                dict(
+                    url_check=dict(
+                        url=url,
+                        status="down",
+                        status_code=500,
```

Comment:
`ClientError` is the base exception and can be raised in multiple cases, so there is no guarantees that the status code is a real 500. It would be better catching `ClientResponseError` 
https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientResponseError

Reply:
If we end up in
```
        except aiohttp.ClientError as e:
```
That means we do not reach the web server at all, due to some unexpected issues, like firewall blocking communication, web server process is dead or defunct, etc. So, we do not get an http error. If I read the ClientResponseError doc it says:

```
exception aiohttp.ClientResponseError[[source]](https://docs.aiohttp.org/en/stable/_modules/aiohttp/client_exceptions.html#ClientResponseError)
These exceptions could happen after we get response from server.
```
Key being "after we get response from server".

If I don't completely misunderstand, I guess the question here is, what do we raise to the user in case we do not get an http error (due to other issues caught in the ClientError exception handling).

Reply:
I simply took 500, as we don't know. We may just as well return 1 or etc?

Reply:
I did some testing, if we end up at
```
        except aiohttp.ClientError as e:
```
AFAIKS, that means we do not reach the web server at all, due to some unexpected issues, like firewall blocking communication, web server process is dead or defunct, ssl issues. IF we get get a response from the web server, that is picked up here:

```
                                    status="up" if resp.status == 200 else "down",
```
And will return what http error we get.

```
exception aiohttp.ClientResponseError[[source]](https://docs.aiohttp.org/en/stable/_modules/aiohttp/client_exceptions.html#ClientResponseError)
These exceptions could happen after we get response from server.
```
Key being "after we get response from server". So, we won't be able to use that.

I set 500, as I didn't know what else to tell the user. We are not getting an http error message, due to use not being able to talk to the http server due to unknown issues. I guess we could return 1, or something non http. 

...This could also be me not understanding what you mean :)

Reply:
If I test against some page which does not exist, we pick up the 404 not found:
```
INFO:ansible_rulebook.rule_generator:calling Web site is down
INFO:ansible_rulebook.engine:call_action run_playbook
INFO:ansible_rulebook.engine:substitute_variables [{'name': 'site_down.yml'}] [{'event': {'url_check': {'status_code': 404, 'url': 'http://rhel9apache.sudo.net/test.html', 'status': 'down'}}, 'fact': {'url_check': {'status_code': 404, 'url': 'http://rhel9apache.sudo.net/test.html', 'status': 'down'}}}]
INFO:ansible_rulebook.engine:action args: {'name': 'site_down.yml'}
INFO:ansible_rulebook.builtin:running Ansible playbook: site_down.yml
INFO:ansible_rulebook.builtin:ruleset: Listen for events on a webhook, rule: Web site is down
INFO:ansible_rulebook.builtin:Calling Ansible runner

PLAY [Site Down] ***************************************************************

TASK [debug] *******************************************************************
ok: [rhel9apache.sudo.net] => {
    "msg": "Website is down!"
}
```

Reply:
Perhaps return "unknown client error"? or simply "client error"

Reply:
You are right, a regular 5xx is returned without exceptions because the client was not initialized with `raise_for_status` param.
In this case I think handling `ClientError` is enough for now. 

I suggest remove the 500 status code and send a new key with the content of the error to be able to use it in the rules conditions.

Reply:
Hope this is what you meant.

Here's how it looks like now. Below scenario is when the webserver is completely down.
I put whatever we got in the exception into the status_code.

```
INFO:ansible_rulebook.rule_generator:calling Web site is down
INFO:ansible_rulebook.engine:call_action run_playbook
INFO:ansible_rulebook.engine:substitute_variables [{'name': 'site_down.yml'}] [{'event': {'url_check': {'status_code': "Cannot connect to host rhel9apache.sudo.net:80 ssl:default [Connect call failed ('192.168.120.130', 80)]", 'url': 'http://rhel9apache.sudo.net', 'status': 'down'}}, 'fact': {'url_check': {'status_code': "Cannot connect to host rhel9apache.sudo.net:80 ssl:default [Connect call failed ('192.168.120.130', 80)]", 'url': 'http://rhel9apache.sudo.net', 'status': 'down'}}}]
```

Reply:
Full output being:
```
[mglantz@darkred event-driven-ansible]$ ansible-rulebook --rulebook website-automation.yml -i inventory --verbose
INFO:ansible_rulebook.app:Starting sources
INFO:ansible_rulebook.app:Starting rules
INFO:ansible_rulebook.engine:run_ruleset
INFO:ansible_rulebook.engine:ruleset define: {"name": "Listen for events on a webhook", "hosts": ["all"], "sources": [{"EventSource": {"name": "ansible.eda.url_check", "source_name": "ansible.eda.url_check", "source_args": {"urls": ["http://rhel9apache.sudo.net"], "delay": 10}, "source_filters": []}}], "rules": [{"Rule": {"name": "Web site is up", "condition": {"AllCondition": [{"EqualsExpression": {"lhs": {"Event": "url_check.status"}, "rhs": {"String": "up"}}}]}, "action": {"Action": {"action": "run_playbook", "action_args": {"name": "site_up.yml"}}}, "enabled": true}}, {"Rule": {"name": "Web site is down", "condition": {"AllCondition": [{"EqualsExpression": {"lhs": {"Event": "url_check.status"}, "rhs": {"String": "down"}}}]}, "action": {"Action": {"action": "run_playbook", "action_args": {"name": "site_down.yml"}}}, "enabled": true}}]}
INFO:ansible_rulebook.engine:load source
INFO:ansible_rulebook.engine:load source filters
INFO:ansible_rulebook.engine:Calling main in ansible.eda.url_check
INFO:ansible_rulebook.engine:Waiting for event from Listen for events on a webhook
INFO:ansible_rulebook.rule_generator:calling Web site is down
INFO:ansible_rulebook.engine:call_action run_playbook
INFO:ansible_rulebook.engine:substitute_variables [{'name': 'site_down.yml'}] [{'event': {'url_check': {'status_code': "Cannot connect to host rhel9apache.sudo.net:80 ssl:default [Connect call failed ('192.168.120.130', 80)]", 'url': 'http://rhel9apache.sudo.net', 'status': 'down'}}, 'fact': {'url_check': {'status_code': "Cannot connect to host rhel9apache.sudo.net:80 ssl:default [Connect call failed ('192.168.120.130', 80)]", 'url': 'http://rhel9apache.sudo.net', 'status': 'down'}}}]
INFO:ansible_rulebook.engine:action args: {'name': 'site_down.yml'}
INFO:ansible_rulebook.builtin:running Ansible playbook: site_down.yml
INFO:ansible_rulebook.builtin:ruleset: Listen for events on a webhook, rule: Web site is down
INFO:ansible_rulebook.builtin:Calling Ansible runner

PLAY [Site Down] ***************************************************************

TASK [debug] *******************************************************************
ok: [rhel9apache.sudo.net] => {
    "msg": "Website is down!"
}

TASK [Restart apache webserver] ************************************************
fatal: [rhel9apache.sudo.net]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: connect to host rhel9apache.sudo.net port 22: No route to host", "unreachable": true}

PLAY RECAP *********************************************************************
rhel9apache.sudo.net       : ok=1    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0   
```

Reply:
If we do have a connection to SSH and it's just either the process which is dead or firewall issue on the host:
```
INFO:ansible_rulebook.rule_generator:calling Web site is down
INFO:ansible_rulebook.engine:call_action run_playbook
INFO:ansible_rulebook.engine:substitute_variables [{'name': 'site_down.yml'}] [{'event': {'url_check': {'status_code': "Cannot connect to host rhel9apache.sudo.net:80 ssl:default [Connect call failed ('192.168.120.130', 80)]", 'url': 'http://rhel9apache.sudo.net', 'status': 'down'}}, 'fact': {'url_check': {'status_code': "Cannot connect to host rhel9apache.sudo.net:80 ssl:default [Connect call failed ('192.168.120.130', 80)]", 'url': 'http://rhel9apache.sudo.net', 'status': 'down'}}}]
INFO:ansible_rulebook.engine:action args: {'name': 'site_down.yml'}
INFO:ansible_rulebook.builtin:running Ansible playbook: site_down.yml
INFO:ansible_rulebook.builtin:ruleset: Listen for events on a webhook, rule: Web site is down
INFO:ansible_rulebook.builtin:Calling Ansible runner

PLAY [Site Down] ***************************************************************

TASK [debug] *******************************************************************
ok: [rhel9apache.sudo.net] => {
    "msg": "Website is down!"
}

TASK [Restart apache webserver] ************************************************
changed: [rhel9apache.sudo.net]

TASK [Permit http traffic in the firewall] *************************************
changed: [rhel9apache.sudo.net]

PLAY RECAP *********************************************************************
rhel9apache.sudo.net       : ok=3    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
200
```